runtime/include/wasm_memory.h

```diff
@@ -0,0 +1,302 @@
+#pragma once
+
+#include <assert.h>
+#include <errno.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+
+#include "types.h" /* PAGE_SIZE */
+#include "wasm_types.h"
+
+#define WASM_MEMORY_MAX (size_t) UINT32_MAX + 1
+
+struct wasm_memory {
+	size_t  size;     /* Initial Size in bytes */
+	size_t  capacity; /* Size backed by actual pages */
+	size_t  max;      /* Soft cap in bytes. Defaults to 4GB */
+	uint8_t data[];
+};
+
+static inline struct wasm_memory *
+wasm_memory_allocate(size_t initial, size_t max)
+{
+	assert(initial > 0);
+	assert(initial <= (size_t)UINT32_MAX + 1);
+	assert(max > 0);
+	assert(max <= (size_t)UINT32_MAX + 1);
+
+	/* Allocate contiguous virtual addresses for struct, full linear memory, and guard page */
+	size_t size_to_alloc = sizeof(struct wasm_memory) + WASM_MEMORY_MAX + /* guard page */ PAGE_SIZE;
+	void * temp          = mmap(NULL, size_to_alloc, PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+	if (temp == MAP_FAILED) {
+		fprintf(stderr, "wasm_memory_allocate - allocation failed, (size: %lu) %s\n", size_to_alloc,
+		        strerror(errno));
+		return NULL;
+	}
+	struct wasm_memory *self = (struct wasm_memory *)temp;
+
+	/* Set the struct and initial pages to read / write */
+	size_t size_to_read_write = sizeof(struct wasm_memory) + initial;
+
+	int rc = mprotect(self, size_to_read_write, PROT_READ | PROT_WRITE);
+	if (rc != 0) {
+		perror("wasm_memory_allocate - prot r/w failed");
+		munmap(self, size_to_alloc);
+		assert(0);
+		return NULL;
+	}
+
+	self->size     = initial;
+	self->capacity = initial;
+	self->max      = max;
+	return self;
+}
+
+static inline void
+wasm_memory_free(struct wasm_memory *self)
+{
+	size_t size_to_free = sizeof(struct wasm_memory) + WASM_MEMORY_MAX + /* guard page */ PAGE_SIZE;
+	munmap(self, size_to_free);
+}
+
+static inline void
+wasm_memory_wipe(struct wasm_memory *self)
+{
+	memset(self->data, 0, self->size);
+}
+
+static inline int
+wasm_memory_expand(struct wasm_memory *self, size_t size_to_expand)
+{
+	size_t target_size = self->size + size_to_expand;
+	if (unlikely(target_size > self->max)) {
+		fprintf(stderr, "wasm_memory_expand - Out of Memory!. %lu out of %lu\n", self->size, self->max);
+		return -1;
+	}
+
+	if (target_size > self->capacity) {
+		int rc = mprotect(self, sizeof(struct wasm_memory) + target_size, PROT_READ | PROT_WRITE);
+		if (rc != 0) {
+			perror("wasm_memory_expand mprotect");
+			return -1;
+		}
+
+		self->capacity = target_size;
+	}
+
+	self->size = target_size;
+	return 0;
+}
+
+/**
+ * Translates WASM offsets into runtime VM pointers
+ * @param offset an offset into the WebAssembly linear memory
+ * @param bounds_check the size of the thing we are pointing to
+ * @return void pointer to something in WebAssembly linear memory
+ */
+static inline void *
+wasm_memory_get_ptr_void(struct wasm_memory *self, uint32_t offset, uint32_t size)
+{
+	assert(offset + size <= self->size);
```

Comment:
We obviously want to change these from `assert` to some form of sandbox error that sledge deals with. Pulling them out into a macro here (to maintain the sloc) might be good? To consider for the futurel.